docs/tutorial/plot_function_transformer.py

```diff
@@ -0,0 +1,193 @@
+    ],
+    remainder="passthrough",
+    verbose_feature_names_out=False,
+)
+mapper.set_output(transform="pandas")
+
+pipe = Pipeline([("mapper", mapper), ("classifier", DecisionTreeClassifier())])
+
+data = DataFrame(
+    [
+        dict(a=1, b=2, f=5),
+        dict(a=4, b=5, f=10),
+    ]
+)
+y = np.array([0, 1], dtype=np.int64)
+pipe.fit(data, y)
+
+try:
+    to_onnx(pipe, data[:1], options={"zipmap": False})
+except Exception as e:
+    print("It does not work:", e)
+
+##################################
+# Use of custom transformer
+# +++++++++++++++++++++++++
+#
+# It is easier to write a custom converter if the FunctionTransformer
+# is implemented as a custom transformer.
+
+
+class OverpriceCalculator(BaseEstimator, TransformerMixin):
+    def __init__(self):
+        pass
+
+    def calculate_overprice(self, x, y):
+        return 100 * (x - y) / y
+
+    def fit(self, X, y=None):
+        return self
+
+    def transform(self, X, y=None):
+        x = X.apply(lambda x: self.calculate_overprice(x.a, x.b), axis=1)
+        return x.values.reshape((-1, 1))
+
+
+mapper = ColumnTransformer(
+    transformers=[
+        ("c", OverpriceCalculator(), ["a", "b"]),
+    ],
+    remainder="passthrough",
+    verbose_feature_names_out=False,
+)
+
+pipe_tr = Pipeline([("mapper", mapper), ("classifier", DecisionTreeClassifier())])
+pipe_tr.fit(data, y)
+
+#############################
+# Both pipelines return the same output.
+assert_allclose(pipe.predict_proba(data), pipe_tr.predict_proba(data))
+
+#############################
+# But the conversion still fails with a different error message.
+
+try:
+    to_onnx(pipe_tr, data[:1], options={"zipmap": False})
+except Exception as e:
+    print("It does not work:", e)
+
+
+#################################
+# Custom converter
+# ++++++++++++++++
+#
+# We need to implement the method `calculate_overprice` in ONNX.
+# The first function returns the expected type and shape.
+
+
+def overprice_shape_calculator(operator):
+    check_input_and_output_numbers(operator, input_count_range=1, output_count_range=1)
+    # Gets the input type, the transformer works on any numerical type.
+    input_type = operator.inputs[0].type.__class__
+    # The first dimension is usually dynamic (batch dimension).
+    input_dim = operator.inputs[0].get_first_dimension()
+    operator.outputs[0].type = input_type([input_dim, 1])
+
+
+def overprice_converter(scope, operator, container):
+    # No need to retrieve the fitted estimator, it is not trained.
+    # op = operator.raw_operator
+    opv = container.target_opset
+    X = operator.inputs[0]
+
+    # 100 * (x-y)/y  --> 100 * (X[0] - X[1]) / X[1]
+
+    zero = np.array([0], dtype=np.int64)
+    one = np.array([1], dtype=np.int64)
+    two = np.array([2], dtype=np.int64)
+    hundred = np.array([100], dtype=np.float32)
+
+    # Slice(data, starts, ends, axes)
+    x0 = OnnxSlice(X, zero, one, one, op_version=opv)
```

Comment:
I feel like the learning curve to convert normal operations to onnx operations is pretty steep. Could you suggest any good tutorials / docs to get more familiar with this?

Reply:
For this example, I chose the API implemented in sklearn-onnx but I could extend the example to make it work with other API you prefer (see [Many ways to implement a custom graph in ONNX](https://sdpython.github.io/doc/onnx-array-api/dev/tutorial/onnx_api.html)) to write onnx graphs.

That said, I understand it is not easy. It takes time to rewrite an existing code with this list of [ONNX Operators](https://onnx.ai/onnx/operators/).

The easiest way is probably to use pytorch but it means to take a dependency on pytorch. The numpy API for onnx is less difficult but I initially started it for small functions. All the other options require to rewrite a function with onnx operators. Most of the operators follow numpy api. A couple of important points: there is no inplace modification allowed in onnx, notation [..., :, ...] can usually be handled with operator Slice but not always, Loop and Tests should be avoided.

I wrote this tutorial about onnx: https://onnx.ai/onnx/intro/python.html.

Reply:
Why is using PyTorch easier than sklearn for converting to onnx?

Reply:
It is not necessary to rewrite your custom code with pytorch. With sklearn, the numpy code must be rewritten. I added a version with onnxscript as well.